### PR TITLE
Replace Lists with Vectors

### LINKED ISSUE #619: Avoid std::list if possible
`std::list` is very useful as a container since it maintains references to objects on insertion / deletion (unlike `std::vector`) however it it has a larger footprint per item in the container, and does not store objects contiguously in memory.  It should be replaced with other containers where possible to do so:

### Current Uses
- [x] `Data1DStore` (as `std::list<std::pair<Data1D, Data1DImportFileFormat>>`)
- [x] `Data2DStore` (as `std::list<std::pair<Data2D, Data2DImportFileFormat>>`)
- [x] `Data3DStore` (as `std::list<std::pair<Data3D, Data3DImportFileFormat>>`)
- [x] `CoreData` (as `std::list<MasterIntra>`) - it appears all `std::list` usages have been replaced with `std::vector`
- [x] `SpeciesAtom` (as `std::list<SpeciesAtom>`) - it appears all `std::list` usages have been replaced with `std::vector`
- [x] Several GUI helper functions (`combopopulator.h`, `tablewidgetupdater.h`, `treewidgetupdater.h`) take `std::list` as data.

### PR DESCRIPTION
**Pull request by IVlaD17**
Closes #619 

Pretty much what's being said in the title. Replacing `std::lists` with `std::vectors` where possible.

Changed the following areas:
- `data1dstore.h`
- `data2dstore.h`
- `data3dstore.h`
- `combopopulator.h`
- `tablewidgetupdater.h`
- `treewidgetupdater.h`

### CODE REVIEW

File: src/classes/data1dstore.cpp [hunk 1]
```diff
@@ -28,18 +28,19 @@ bool Data1DStore::addData(std::string_view dataName, LineParser &parser, int sta
 // Check to see if the named data is present in the store
 bool Data1DStore::containsData(std::string_view name) const
 {
-    return std::find_if(data_.begin(), data_.end(), [&name](auto &data) { return data.first.tag() == name; }) != data_.end();
+    return std::find_if(data_.begin()->get(), data_.end()->get(), [&name](auto &data) { return data.first.tag() == name; }) !=
+           data_.end()->get();
```
Comment (trisyoungs) [inline src/classes/data1dstore.cpp:32]: This doesn't look right and wouldn't traverse the vector:
```suggestion
    return std::find_if(data_.begin(), data_.end(), [&name](auto &data) { return data->first.tag() == name; }) !=
           data_.end();
```

File: src/classes/data1dstore.cpp [hunk 2]
```diff
@@ -28,18 +28,19 @@ bool Data1DStore::addData(std::string_view dataName, LineParser &parser, int sta
...
 }
 
 // Return named data
 OptionalReferenceWrapper<const Data1D> Data1DStore::data(std::string_view name) const
 {
-    auto it = std::find_if(data_.begin(), data_.end(), [&name](auto &data) { return data.first.tag() == name; });
-    if (it == data_.end())
+    auto it = std::find_if(data_.begin()->get(), data_.end()->get(), [&name](auto &data) { return data.first.tag() == name; });
+    if (it == data_.end()->get())
```
Comment (trisyoungs) [inline src/classes/data1dstore.cpp:39]: Same here:
```suggestion
    auto it = std::find_if(data_.begin(), data_.end(), [&name](auto &data) { return data->first.tag() == name; });
    if (it == data_.end())
```

File: src/classes/data2dstore.cpp [hunk 3]
```diff
@@ -28,18 +28,19 @@ bool Data2DStore::addData(std::string_view dataName, LineParser &parser, int sta
 // Check to see if the named data is present in the store
 bool Data2DStore::containsData(std::string_view name) const
 {
-    return std::find_if(data_.begin(), data_.end(), [&name](auto &data) { return data.first.tag() == name; }) != data_.end();
+    return std::find_if(data_.begin()->get(), data_.end()->get(), [&name](auto &data) { return data.first.tag() == name; }) !=
+           data_.end()->get();
```
Comment (trisyoungs) [inline src/classes/data2dstore.cpp:32]:
```suggestion
    auto it = std::find_if(data_.begin(), data_.end(), [&name](auto &data) { return data->first.tag() == name; });
    if (it == data_.end())
```

File: src/classes/data2dstore.cpp [hunk 4]
```diff
@@ -28,18 +28,19 @@ bool Data2DStore::addData(std::string_view dataName, LineParser &parser, int sta
...
 }
 
 // Return named data
 OptionalReferenceWrapper<const Data2D> Data2DStore::data(std::string_view name) const
 {
-    auto it = std::find_if(data_.begin(), data_.end(), [&name](auto &data) { return data.first.tag() == name; });
-    if (it == data_.end())
+    auto it = std::find_if(data_.begin()->get(), data_.end()->get(), [&name](auto &data) { return data.first.tag() == name; });
+    if (it == data_.end()->get())
```
Comment (trisyoungs) [inline src/classes/data2dstore.cpp:39]:
```suggestion
    auto it = std::find_if(data_.begin(), data_.end(), [&name](auto &data) { return data->first.tag() == name; });
    if (it == data_.end())
```

File: src/classes/data3dstore.cpp [hunk 5]
```diff
@@ -28,18 +28,19 @@ bool Data3DStore::addData(std::string_view dataName, LineParser &parser, int sta
...
+           data_.end()->get();
 }
 
 // Return named data
 OptionalReferenceWrapper<const Data3D> Data3DStore::data(std::string_view name) const
 {
-    auto it = std::find_if(data_.begin(), data_.end(), [&name](auto &data) { return data.first.tag() == name; });
-    if (it == data_.end())
+    auto it = std::find_if(data_.begin()->get(), data_.end()->get(), [&name](auto &data) { return data.first.tag() == name; });
+    if (it == data_.end()->get())
```
Comment (trisyoungs) [inline src/classes/data3dstore.cpp:39]:
```suggestion
    auto it = std::find_if(data_.begin(), data_.end(), [&name](auto &data) { return data->first.tag() == name; });
    if (it == data_.end())
```

File: src/keywords/data1dstore.cpp [hunk 6]
```diff
@@ -45,8 +45,9 @@ bool Data1DStoreKeyword::deserialise(LineParser &parser, int startArg, const Cor
 // Serialise data to specified LineParser
 bool Data1DStoreKeyword::serialise(LineParser &parser, std::string_view keywordName, std::string_view prefix) const
 {
-    for (const auto &[data, format] : data_.data())
+    for (const auto sharedDataPointer : data_.data())
```
Comment (trisyoungs) [inline src/keywords/data1dstore.cpp:48]: Just a small thing, but we can avoid copying the shared pointer here and use a reference:
```suggestion
    for (const auto &sharedDataPointer : data_.data())
```

File: src/keywords/data2dstore.cpp [hunk 7]
```diff
@@ -40,8 +40,9 @@ bool Data2DStoreKeyword::deserialise(LineParser &parser, int startArg, const Cor
 // Serialise data to specified LineParser
 bool Data2DStoreKeyword::serialise(LineParser &parser, std::string_view keywordName, std::string_view prefix) const
 {
-    for (const auto &[data, format] : data_.data())
+    for (const auto sharedDataPointer : data_.data())
```
Comment (trisyoungs) [inline src/keywords/data2dstore.cpp:43]:
```suggestion
    for (const auto &sharedDataPointer : data_.data())
```

File: src/modules/datatest/process.cpp [hunk 8]
```diff
@@ -22,8 +22,9 @@ bool DataTestModule::process(Dissolve &dissolve, ProcessPool &procPool)
     Messenger::print("\n");
 
     // Loop over reference one-dimensional data supplied
-    for (auto &[referenceData, format] : test1DData_.data())
+    for (auto sharedDataPointer : test1DData_.data())
```
Comment (trisyoungs) [inline src/modules/datatest/process.cpp:25]:
```suggestion
    for (auto &sharedDataPointer : test1DData_.data())
```

File: src/modules/datatest/process.cpp [hunk 9]
```diff
@@ -63,8 +64,9 @@ bool DataTestModule::process(Dissolve &dissolve, ProcessPool &procPool)
     }
 
     // Loop over reference two-dimensional data supplied
-    for (auto &[referenceData, format] : test2DData_.data())
+    for (auto sharedDataPointer : test2DData_.data())
```
Comment (trisyoungs) [inline src/modules/datatest/process.cpp:67]:
```suggestion
    for (auto &sharedDataPointer : test2DData_.data())
```

File: src/modules/rdf/functions.cpp [hunk 10]
```diff
@@ -645,8 +645,9 @@ bool RDFModule::testReferencePartials(const Data1DStore &testData, double testTh
     LineParser parser;
 
     // Loop over supplied test data and see if we can locate it amongst our PartialSets
-    for (auto &[data, format] : testData.data())
+    for (auto sharedDataPointer : testData.data())
```
Comment (trisyoungs) [inline src/modules/rdf/functions.cpp:648]:
```suggestion
    for (auto &sharedDataPointer : testData.data())
```

**Review by trisyoungs — CHANGES_REQUESTED** [review]
Looks good.  Some comments over a few of the loops which should be investigated.

**Comment by rprospero** [comment]
I agree with Tristan's comments, but otherwise approve.

**Review by trisyoungs — APPROVED** [review]

**Pull request merged**